pkg/ebpf/dns_snooper.go

```diff
@@ -87,129 +82,149 @@ func (s *SocketFilterSnooper) Resolve(connections []ConnectionStats) map[util.Ad
 
 func (s *SocketFilterSnooper) GetStats() map[string]int64 {
 	stats := s.cache.Stats()
-	stats["packets_captured"] = atomic.SwapInt64(&s.packets, 0)
+	stats["socket_polls"] = atomic.SwapInt64(&s.polls, 0)
+	stats["packets_processed"] = atomic.SwapInt64(&s.processed, 0)
+	stats["packets_captured"] = atomic.SwapInt64(&s.captured, 0)
+	stats["packets_dropped"] = atomic.SwapInt64(&s.dropped, 0)
+	stats["decoding_errors"] = atomic.SwapInt64(&s.decodingErrors, 0)
+
 	return stats
 }
 
 // Close terminates the DNS traffic snooper as well as the underlying socket and the attached filter
 func (s *SocketFilterSnooper) Close() {
 	close(s.exit)
 	s.wg.Wait()
+	s.source.Close()
+	s.cache.Close()
+}
 
-	if err := bpflib.DetachSocketFilter(s.socketFilter, s.socketFD); err != nil {
-		log.Errorf("error detaching socket filter: %s", err)
+// processPacket retrieves DNS information from the received packet data and adds it to
+// the reverse DNS cache. The underlying packet data can't be referenced after this method
+// call since gopacket re-uses it.
+func (s *SocketFilterSnooper) processPacket(data []byte) {
+	translation := s.parser.Parse(data)
+	if translation == nil {
+		atomic.AddInt64(&s.decodingErrors, 1)
+		return
 	}
 
-	s.tpacket.Close()
-	s.cache.Close()
+	s.cache.Add(translation, time.Now())
 }
 
-func (s *SocketFilterSnooper) run(packets <-chan gopacket.Packet) {
-	for packet := range packets {
-		layer := packet.Layer(layers.LayerTypeDNS)
-		if layer == nil {
+func (s *SocketFilterSnooper) pollPackets() {
+	for {
+		data, _, err := s.source.ZeroCopyReadPacketData()
+
+		// Properly synchronizes termination process
+		select {
+		case <-s.exit:
+			return
+		default:
+		}
+
+		if err == nil {
+			s.processPacket(data)
 			continue
 		}
-		dns, ok := layer.(*layers.DNS)
-		if !ok {
+
+		// Immediately retry for temporary network errors
+		if nerr, ok := err.(net.Error); ok && nerr.Temporary() {
```

Comment:
Isn't this reading from shared memory?  What is the network error that would occur here?

Reply:
good point. I should revisit that code now that I changed things on the socket polling.